qiskit/qasm/qasmparser.py

```diff
@@ -373,16 +373,10 @@ def p_format(self, program):
         """
         format : FORMAT
         """
-        program[0] = node.Format(program[1])
-
-    def p_format_0(self, _):
-        """
-        format : FORMAT error
-        """
-        version = "2.0;"
-        raise QasmError(
-            "Invalid version string. Expected '" + version + "'.  Is the semicolon missing?"
-        )
+        version = node.Format(program[1])
+        if (version.majorversion != "2") or (version.minorversion != "0"):
+            raise QasmError(str("Invalid version string. This module supports OPENQASM 2.0; only."))
+        program[0] = version
```

Comment:
This new form works rather well, thanks!  Can we put in the parsed version token (e.g. "Invalid version string: ... This module only supports OpenQASM 2") too?  Also, just to be a little more permissive - while OpenQASM 2 techncially only allows the version to be `2.0`, OpenQASM 3 currently allows it to be a floating point number (though in practice we ought to change that to something more suited to representing versions).  That means that `OPENQASM 3;` should be a valid header for OpenQASM 3 programs, but it'll throw the old
```python
QasmError: 'Unable to match any token rule, got -->O<-- Check your OPENQASM source and any include statements.'
```
error here.  We could relax the parser rule to `OPENQASM\s+[0-9]\+(\.[0-9]\+)?'` here, and default to `minor_version = 0` if we don't match that group.  (We'll need tests for `OPENQASM 3;` throwing, and `OPENQASM 2;` being accepted)

I think that's as far as I'd want to go - in theory it might be possible to break the parsing with a super weird error message if we put in a _token_ that OpenQASM 3 has, but 2 doesn't (like funny Unicode stuff), but off a very brief attempt, that seemed to work fine for me here.

Reply:
Got it, makes sense. Pushing this now.